cdap-data-fabric/src/main/java/co/cask/cdap/data/format/CombinedLogRecordFormat.java

```diff
@@ -0,0 +1,149 @@
+    String bodyAsStr = Bytes.toString(event.getBody(), charset);
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    List<String> parts = getLogEntries(bodyAsStr);
+    List<Schema.Field> fields = schema.getFields();
+    int index = 0;
+    while (index < fields.size()) {
+      Schema.Field field = fields.get(index);
+      String val = (parts.size() < index || (parts.get(index).equals("-") &&
+                                             field.getSchema().getType() != Schema.Type.STRING))
+                    ? null : parts.get(index);
+      builder.convertAndSet(fields.get(index).getName(), val);
+      index++;
+    }
+
+    return builder.build();
+  }
+
+
+  @Override
+  protected Schema getDefaultSchema() {
+    return Schema.recordOf("streamEvent",
+           Schema.Field.of("remote_host", Schema.unionOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("remote_login ", Schema.unionOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("auth_user", Schema.unionOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("date", Schema.unionOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("request", Schema.unionOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("status", Schema.unionOf(Schema.of(Schema.Type.INT), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("content_length", Schema.unionOf(Schema.of(Schema.Type.INT), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("referrer", Schema.unionOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("user_agent", Schema.unionOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.NULL))));
+  }
+
+
+  @Override
+  protected void validateSchema(Schema desiredSchema) throws UnsupportedTypeException {
+    // a valid schema is a record of simple types.
+    Iterator<Schema.Field> fields = desiredSchema.getFields().iterator();
+    while (fields.hasNext()) {
+      Schema.Field field = fields.next();
+      Schema schema = field.getSchema();
+      boolean isSimple = schema.getType().isSimpleType();
+      boolean isNullableSimple = schema.isNullableSimple();
+      if (!isSimple && !isNullableSimple) {
+        throw new UnsupportedTypeException("Field " + field.getName() + " is of invalid type.");
+      }
+    }
+  }
+
+  // parse CLF logEvent and get the record values.
+  private List<String> getLogEntries(String logEvent) {
+    List<String> parts = Lists.newArrayList();
+    boolean  done = false;
+    int start = 0;
+    while (!done) {
+      if (logEvent.charAt(start) == ' ') {
+        // Skip empty spaces
+        start++;
+      } else {
+        start = addNextLogEntry(logEvent, start, parts);
+      }
+
+      if (start >= logEvent.length()) {
```

Comment:
You can move this condition as the while loop condition `while (start < logEvent.length())`